evaluation/scripts/evaluate.py

```diff
@@ -0,0 +1,363 @@
+
+    ax1.bar(metric_list, pass_rate)
+
+    ax1.set_ylabel("passing rate")
+    ax1.set_title("Passing rate of evaluation metrics")
+    plt.savefig("passing_rate.png")
+    plt.close(fig)
+
+    fig, ax2 = plt.subplots()
+    # bar_labels = ['red', 'blue', '_red', 'orange']
+    # bar_colors = ['tab:red', 'tab:blue', 'tab:red', 'tab:orange']
+
+    ax2.bar(metric_list, mean_rate)
+
+    ax2.set_ylabel("mean score")
+    ax2.set_title("Mean score of evaluation metrics")
+    plt.savefig("mean_score.png")
+    plt.close(fig)
+
+    penguin_means = {
+        "Max": tuple(max_list),
+        "Min": tuple(min_list),
+        "Mean": tuple(mean_list),
+    }
+
+    x = np.arange(len(metric_name))  # the label locations
+    width = 0.25  # the width of the bars
+    multiplier = 0
+
+    fig, ax3 = plt.subplots(layout="constrained")
+
+    for attribute, measurement in penguin_means.items():
+        offset = width * multiplier
+        rects = ax3.bar(x + offset, measurement, width, label=attribute)
+        ax3.bar_label(rects, padding=3)
+        multiplier += 1
+
+    # Add some text for labels, title and custom x-axis tick labels, etc.
+    ax3.set_title("Evaluation results")
+    ax3.set_xticks(x + width, tuple(metric_name))
+    ax3.legend(loc="upper left", ncols=3)
+    ax3.set_ylim(0, 250)
+
+    plt.savefig("eval.png")
+    plt.close(fig)
+
+    # summary statistics
+    with open(results_dir / "summary.json", "w", encoding="utf-8") as summary_file:
+        summary_file.write(json.dumps(summary, indent=4))
+
+    with open(results_dir / "evaluate_parameters.json", "w", encoding="utf-8") as parameters_file:
+        parameters = {
+            "evaluation_gpt_model": openai_config.model,
+            "evaluation_timestamp": int(time.time()),
+            "testdata_path": str(testdata_path),
+            "target_url": target_url,
+            "target_parameters": target_parameters,
+            "num_questions": num_questions,
+        }
+        parameters_file.write(json.dumps(parameters, indent=4))
+    logger.info("Evaluation results saved in %s", results_dir)
+    return True
+
+
+def process_config(obj: dict):
+    """Replace special markers in a config dict with their values:
+    * <TIMESTAMP> with current timestamp
+    * <READFILE> with contents of file
+    """
+    if isinstance(obj, dict):
+        for key in obj:
+            if isinstance(obj[key], dict):
+                process_config(obj[key])
+            elif isinstance(obj[key], str) and "<TIMESTAMP>" in obj[key]:
+                logger.info("Replaced %s in config with timestamp", key)
+                obj[key] = obj[key].replace("<TIMESTAMP>", str(int(time.time())))
+            elif isinstance(obj[key], str) and "<READFILE>" in obj[key]:
+                with open(obj[key].replace("<READFILE>", ""), encoding="utf-8") as f:
+                    logger.info("Replaced %s in config with contents of %s", key, f.name)
+                    obj[key] = f.read()
```

Comment:
What's the purpose of this? Why do we need to replace the placeholder? Why can't we just generate the name of the experiments directory with the timestamp and attach it to the `results_dir` specified in the config, which will simply be `results`? And what about the readfile? Are we even using it?